compiler/rustc_mir_build/src/check_tail_calls.rs

```diff
@@ -0,0 +1,382 @@
+        if self.is_closure {
+            self.report_in_closure(expr);
+            return;
+        }
+
+        let BodyTy::Fn(caller_sig) = self.thir.body_type
+            else { span_bug!(call.span, "`become` outside of functions should have been disallowed by hit_typeck") };
+
+        let ExprKind::Scope { value, .. } = call.kind
+            else { span_bug!(call.span, "expected scope, found: {call:?}") };
+        let value = &self.thir[value];
+
+        if matches!(
+            value.kind,
+            ExprKind::Binary { .. }
+                | ExprKind::Unary { .. }
+                | ExprKind::AssignOp { .. }
+                | ExprKind::Index { .. }
+        ) {
+            self.report_builtin_op(call, expr);
+            return;
+        }
+
+        let ExprKind::Call { ty, fun, ref args, from_hir_call, fn_span } = value.kind
+            else {
+                self.report_non_call(value, expr);
+                return
+            };
+
+        if !from_hir_call {
+            self.report_op(ty, args, fn_span, expr);
+        }
+
+        // Closures in thir look something akin to
+        // `for<'a> extern "rust-call" fn(&'a [closure@...], ()) -> <[closure@...] as FnOnce<()>>::Output {<[closure@...] as Fn<()>>::call}`
+        // So we have to check for them in this weird way...
+        if let &ty::FnDef(did, substs) = ty.kind() {
+            let parent = self.tcx.parent(did);
+            let fn_ = self.tcx.require_lang_item(LangItem::Fn, Some(expr.span));
+            let fn_once = self.tcx.require_lang_item(LangItem::FnOnce, Some(expr.span));
+            let fn_mut = self.tcx.require_lang_item(LangItem::FnMut, Some(expr.span));
+            if [fn_, fn_once, fn_mut].contains(&parent) {
+                if substs.first().and_then(|arg| arg.as_type()).is_some_and(|t| t.is_closure()) {
+                    self.report_calling_closure(
+                        &self.thir[fun],
+                        substs[1].as_type().unwrap(),
+                        expr,
+                    );
+
+                    // Tail calling is likely to cause unrelated errors (ABI, argument mismatches)
+                    return;
+                }
+            };
+        }
+
+        // Erase regions since tail calls don't care about lifetimes
+        let callee_sig =
+            self.tcx.normalize_erasing_late_bound_regions(self.param_env, ty.fn_sig(self.tcx));
+
+        if caller_sig.abi != callee_sig.abi {
+            self.report_abi_mismatch(expr.span, caller_sig.abi, callee_sig.abi);
+        }
+
+        if caller_sig.inputs_and_output != callee_sig.inputs_and_output {
+            if caller_sig.inputs() != callee_sig.inputs() {
+                self.report_arguments_mismatch(expr.span, caller_sig, callee_sig);
+            }
+
+            if caller_sig.output() != callee_sig.output() {
+                span_bug!(expr.span, "hir typeck should have checked the return type already");
+            }
```

Comment:
This panics w/ opaque types, for example:
```rust
fn f() -> impl Sized {
    become g();
}

fn g() {}
```
or
```rust
fn f() -> impl std::future::Future<Output = ()> {
    become g();
}

async fn g() {}
```

both of those should probably work